vyper/venom/passes/branch_optimization.py

```diff
@@ -0,0 +1,32 @@
+from vyper.venom.analysis.cfg import CFGAnalysis
+from vyper.venom.analysis.dfg import DFGAnalysis
+from vyper.venom.passes.base_pass import IRPass
+
+
+class BranchOptimizationPass(IRPass):
+    """
+    This pass optimizes branches inverting jnz instructions where appropriate
+    """
+
+    def _optimize_branches(self) -> None:
+        fn = self.function
+        for bb in fn.get_basic_blocks():
+            term_inst = bb.instructions[-1]
+            if term_inst.opcode != "jnz":
+                continue
+
+            prev_inst = self.dfg.get_producing_instruction(term_inst.operands[0])
```

Comment:
no recursion needed? could be a chain of iszeros

Reply:
I would prefer this to be part of a different pass, but I have implemented it right into this one (made no change in cases handled)